src/vs/vscode.d.ts

```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
 	}
 
 	/**
+	 * A node provider for the tree explorer contributed by extension.
+	 *
+	 * Providers are registered through (#workspace.registerTreeExplorerNodeProvider) with a
+	 * `providerId` that corresponds to the `treeExplorerNodeProviderId` in the extension's
+	 * `contributes.explorer` section.
+	 *
+	 * The contributed tree explorer will ask the corresponding provider to provide the root
+	 * node and resolve children for each node. In addition, the provider could **optionally**
+	 * provide the following information for each node:
+	 * - label: A human-readable label used for rendering the node.
+	 * - hasChildren: Whether the node has children and is expandable.
+	 * - clickCommand: A command to execute when the node is clicked.
+	 */
+	export interface TreeExplorerNodeProvider<T> {
```

Comment:
We actually had it earlier as you're requesting and @kieferrm convinced us to go in this direction.

> Your sample proofs that the label provider indirection is artificial as the actual data is stored in the node and not computed by a provider

There are a few benefits to doing it this way:

- Instead of passing in some TreeItem data structure, the functions work on the actual backing object itself. This removes the need to go and construct a bunch of node objects but instead derive the label from the object you're working on. This extends to the click command, has children, etc. 
- In the previous paradigm the TreeItem typically required some `data` attribute to disambiguous duplicate labels that appear at multiple places within the tree. Consider a class explorer that have multiple nested objects that may have the same labels but has different click, children etc, 
  
  ```
  class foo
    int a
    class foo
      int b
  ```

  When it comes to provide the second `foo`'s nodes, you need some additional context to handle this. This came up in the examples @octref wrote, so it would be typical to add a `TreeItem.data` property anyway, this just cuts the middle man.
- It becomes very easy to switch all labels to use a different function by simply changing the function and triggering an update for example. 

There's some more context on this in https://github.com/octref/vscode/issues/4 and https://github.com/octref/vscode/issues/5.

I was also originally opposed to this way but it's definitely grown on me and I think it's quite a nice way to work on a tree structure. I agree consistency very important - just some background/insights into the current situation.